### DISCUSSION TITLE
Simple example for demonstrating equivariance (YR(g)r) = DY(r))

### DISCUSSION THREAD

**Discussion by ipcamit** [e3nn/e3nn#345] [Q&A] [answered]
I am trying to get my head around E3NN and Equivarient networks in general using tensor products (At present I am using Satorras et al.  kind of architecture). As per Section 4.1.1 of Tensor Field networks [paper](https://arxiv.org/abs/1802.08219) Equivariance is defined as

$$
Y^l_m (R(g)\hat{r}) = \sum_{m′} D^l_{mm′}(g) Y^l_{m′} (\hat{r}) 
$$

But when I try and visualize it using e3nn (or other libraries like `spherical`) I do not get expected results. For example given the code below should not `YRx` and `DYx` be same? What am I doing wrong here? Should the output not be zero?

```python
import torch
import e3nn

x1 = torch.tensor([1.0, 0.0, 0.0])

# Generate Real Spherical Harmonics L=2
K1 = 0.5 * torch.sqrt(torch.tensor(15./torch.pi))
K2 = 0.25 * torch.sqrt(torch.tensor(5./torch.pi))
x,y,z = x1; r2 = torch.sum(x1**2)
Yx = torch.tensor([
    K1 * x*y/r2, 
    K1 * y*z/r2, 
    K2 * (3*z**2 - r2)/r2, 
    K1 * x*z/r2, 
    K1 * (x**2 - y**2)/r2]
    )

# Random unitary matrix
R,Res = torch.linalg.qr(torch.rand(3,3))

# Rotated coordinates SH
Rx1 = R @ x1
x,y,z = Rx1; r2 = torch.sum(Rx1**2)
YRx = torch.tensor([
    K1 * x*y/r2, 
    K1 * y*z/r2, 
    K2 * (3*z**2 - r2)/r2, 
    K1 * x*z/r2, 
    K1 * (x**2 - y**2)/r2
    ])

# Generate Wigner D matrix (This and e3nn.o3.wigner_D gives same results)
ir = e3nn.o3.Irreps("1x2e")
D = ir.D_from_matrix(torch.transpose(R,0,1))

# ∑𝑚′𝐷(𝑙)𝑚𝑚′(𝑔)𝑌(𝑙)𝑚′(𝑟̂ )
DYx = D @ Yx

print(DYx - YRx)
```

**Comment by mariogeiger** [comment, marked as answer]
I spotted 3 problems:
- we use y in the middle
- 0.5 was missing
- transpose of R

```
x,y,z = z,x,y
Yx = torch.tensor([
    K1 * x*y/r2, 
    K1 * y*z/r2, 
    K2 * (3*z**2 - r2)/r2, 
    K1 * x*z/r2, 
    0.5 * K1 * (x**2 - y**2)/r2]
    )
```

or alternatively
```
Yx = e3nn.o3.spherical_harmonics([2], x1, normalize=True)
```

Also you have to do `D = ir.D_from_matrix(R)` instead of `D = ir.D_from_matrix(torch.transpose(R,0,1))`

**Reply by ipcamit** [reply]
Thank you for the reply. It works now. I missed the 0.5 when copying the coefficients. The y in the middle is bit interesting, any particular reason for that convention?

**Reply by mariogeiger** [reply]
It's to have 
- the spherical harmonics L=1 equals to the identity 
- R equal to D for L=1
